hoodie-client/src/main/java/com/uber/hoodie/table/BulkInsertPartitioner.java

```diff
@@ -0,0 +1,192 @@
+    private static Logger logger = LogManager.getLogger(BulkInsertPartitioner.class);
+    private static final int BUCKET_MULTIPLIER = 4;
+    private final int expectedOutputSparkPartitions;
+    private final int numBuckets;
+    private HoodiePartitionMapper hoodiePartitionMapper;
+    private int numOfPartitions = -1;
+
+    private BulkInsertPartitioner(int outputSparkPartitions) {
+        expectedOutputSparkPartitions = Math.max(1, outputSparkPartitions);
+        numBuckets = expectedOutputSparkPartitions * BUCKET_MULTIPLIER;
+        this.hoodiePartitionMapper = new HoodiePartitionMapper();
+    }
+
+    private void init(JavaRDD<HoodieRecord<T>> records) {
+        TreeMap<String, Long> bucketRecordCount = mapRecordsToBucket(records);
+        Counter totalRecordCount = new Counter(0);
+        // Count total records.
+        bucketRecordCount.values().forEach(v -> totalRecordCount.addCount(v));
+        long expectedRecordsPerSparkPartition = totalRecordCount.getCount() / expectedOutputSparkPartitions;
+        Map<String, Counter> currHoodiePartitionRecordCnt = new HashMap<>();
+        for (Entry<String, Long> entry : bucketRecordCount.entrySet()) {
+            String[] parts = entry.getKey().split(" ");
+            String hoodiePartition = parts[0];
+            if (!currHoodiePartitionRecordCnt.containsKey(hoodiePartition)) {
+                currHoodiePartitionRecordCnt.put(hoodiePartition, new Counter(0));
+                numOfPartitions++;
+            }
+            Counter cnt = currHoodiePartitionRecordCnt.get(hoodiePartition);
+            if (cnt.getCount() >= expectedRecordsPerSparkPartition) {
+                // Reset count and also change output spark partition. That way going forward records for same hoodie
+                // partition will be sent to different spark output partition.
+                cnt.setCount(0);
+                numOfPartitions++;
+            } else {
+                cnt.addCount(entry.getValue());
+            }
+            hoodiePartitionMapper.add(entry.getKey(), numOfPartitions);
+        }
+    }
+
+    @Override
+    public int getPartition(final Object key) {
+        Tuple2<HoodieRecord<T>, Integer> pair = (Tuple2<HoodieRecord<T>, Integer>) key;
+        return hoodiePartitionMapper.get(getBucketKey(pair._1(), pair._2()));
+    }
+
+    @Override
+    public int numPartitions() {
+        return numOfPartitions + 1;
+    }
+
+    public static <T extends HoodieRecordPayload> JavaRDD<HoodieRecord<T>> repartitionRecords(
+        JavaRDD<HoodieRecord<T>> records, int outputSparkPartitions) {
+        // As we are going over input records twice; we need to cache input records.
+        JavaRDD<HoodieRecord<T>> cachedRecords = records.persist(StorageLevel.MEMORY_AND_DISK_SER());
+        BulkInsertPartitioner<T> bulkInsertPartitioner = new BulkInsertPartitioner<>(outputSparkPartitions);
+        bulkInsertPartitioner.init(records);
```

Comment:
lets pass in the records into the constructor?  is there a specific reason for the init method?